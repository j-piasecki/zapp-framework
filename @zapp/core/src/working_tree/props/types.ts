export interface Padding {
  top: number
  bottom: number
  start: number
  end: number
}

export enum PointerEventType {
  DOWN,
  UP,
  MOVE,
  ENTER,
  LEAVE,
}

export interface PointerData {
  target: string
  id: number
  x: number
  y: number
  timestamp: number
  type: PointerEventType
  capture: () => void
}

export interface ConfigType {
  id: string

  weight?: number
  background?: number
  cornerRadius?: number

  borderWidth?: number
  borderColor?: number

  lineWidth?: number
  startAngle?: number
  endAngle?: number

  width?: number
  fillWidth?: number

  height?: number
  fillHeight?: number

  padding?: Padding
  offsetX?: number
  offsetY?: number

  text?: string
  textSize?: number
  textColor?: number

  arrangement?: Arrangement
  alignment?: Alignment
  stackAlignment?: StackAlignment

  source?: string
  innerOffsetX?: number
  innerOffsetY?: number
  originX?: number
  originY?: number
  rotation?: number

  onPointerDown?: (event: PointerData) => void
  onPointerMove?: (event: PointerData) => void
  onPointerUp?: (event: PointerData) => void
  onPointerEnter?: (event: PointerData) => void
  onPointerLeave?: (event: PointerData) => void
}

/**
 * Positioning along the main axis
 */
export enum Arrangement {
  SpaceBetween,
  SpaceAround,
  SpaceEvenly,
  Start,
  Center,
  End,
}

/**
 * Positioning along the secondary axis
 */
export enum Alignment {
  Start,
  Center,
  End,
}

export enum StackAlignment {
  CenterStart,
  Center,
  CenterEnd,
  TopStart,
  TopCenter,
  TopEnd,
  BottomStart,
  BottomCenter,
  BottomEnd,
}
