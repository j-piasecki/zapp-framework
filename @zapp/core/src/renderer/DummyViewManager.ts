import { RenderNode } from './Renderer.js'
import { ViewManager } from './ViewManager.js'

export class DummyViewManager extends ViewManager {
  private nextViewId = 0

  get screenWidth() {
    return 400
  }

  get screenHeight() {
    return 400
  }

  public createView(node: RenderNode) {
    console.log('create', node.id)

    return this.nextViewId++
  }

  public dropView(node: RenderNode) {
    console.log('drop', node.id)
  }

  public updateView(previous: RenderNode, next: RenderNode) {
    console.log('update', next.id)
  }

  public getScrollOffset(): { x: number; y: number } {
    return { x: 0, y: 0 }
  }

  public measureText(
    text: string,
    _node: RenderNode,
    _availableWidth: number,
    _availableHeight: number
  ): { width: number; height: number } {
    return {
      width: text.length,
      height: 1,
    }
  }

  public isRTL(): boolean {
    return false
  }
}
