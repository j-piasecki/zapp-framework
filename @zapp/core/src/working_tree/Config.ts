export class Config {
  public id: string

  constructor(id: string) {
    this.id = id
  }
}
