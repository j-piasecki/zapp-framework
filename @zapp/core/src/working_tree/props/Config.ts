export function Config(id: string) {
  return new ConfigBuilder(id)
}

export interface Padding {
  top: number
  bottom: number
  start: number
  end: number
}

export interface ConfigType {
  id: string

  fillSize?: boolean

  weight?: number
  background?: number
  cornerRadius?: number

  width?: number
  fillWidth?: number

  height?: number
  fillHeight?: number

  padding?: Padding

  text?: string

  // props that may be inherited from the parent
  textSize?: number
  textColor?: number
}

export class ConfigBuilder {
  private config: ConfigType

  constructor(id: string) {
    this.config = {
      id: id,
    }
  }

  public build() {
    return this.config
  }

  public fillSize(): Omit<this, 'width' | 'fillWidth' | 'height' | 'fillHeight' | 'fillSize'> {
    this.config.fillSize = true
    return this
  }

  public fillWidth(portion: number): Omit<this, 'width' | 'fillWidth' | 'fillSize'> {
    this.config.fillWidth = portion
    return this
  }

  public fillHeight(portion: number): Omit<this, 'height' | 'fillHeight' | 'fillSize'> {
    this.config.fillHeight = portion
    return this
  }

  public width(width: number): Omit<this, 'width' | 'fillWidth' | 'fillSize'> {
    this.config.width = width
    return this
  }

  public height(height: number): Omit<this, 'height' | 'fillHeight' | 'fillSize'> {
    this.config.height = height
    return this
  }

  public weight(weight: number): Omit<this, 'weight'> {
    this.config.weight = weight
    return this
  }

  public padding(padding: number): Omit<this, 'padding'>
  public padding(vertical: number, horizontal: number): Omit<this, 'padding'>
  public padding(start: number, top: number, end: number, bottom: number): Omit<this, 'padding'>
  public padding(start: number, top?: number, end?: number, bottom?: number): Omit<this, 'padding'> {
    if (top !== undefined && end !== undefined && bottom !== undefined) {
      this.config.padding = {
        start: start,
        top: top,
        end: end,
        bottom: bottom,
      }
    } else if (top !== undefined) {
      this.config.padding = {
        start: top,
        top: start,
        end: top,
        bottom: start,
      }
    } else {
      this.config.padding = {
        start: start,
        top: start,
        end: start,
        bottom: start,
      }
    }

    return this
  }

  public background(background: number): Omit<this, 'background'> {
    this.config.background = background
    return this
  }

  public cornerRadius(radius: number): Omit<this, 'cornerRadius'> {
    this.config.cornerRadius = radius
    return this
  }

  public textColor(textColor: number): Omit<this, 'textColor'> {
    this.config.textColor = textColor
    return this
  }

  public textSize(textSize: number): Omit<this, 'cornerRatextSizedius'> {
    this.config.textSize = textSize
    return this
  }
}
