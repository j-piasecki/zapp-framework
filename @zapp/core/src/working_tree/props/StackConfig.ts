import { ConfigBuilder } from './Config.js'
import { StackAlignment } from './types.js'

export function StackConfig(id: string) {
  return new StackConfigBuilder(id)
}

export class StackConfigBuilder extends ConfigBuilder {
  public alignment(alignment: StackAlignment) {
    this.config.stackAlignment = alignment
    return this
  }
}
