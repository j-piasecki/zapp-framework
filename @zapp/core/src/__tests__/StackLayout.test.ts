import { WorkingTree } from '../working_tree/WorkingTree'
import { Stack } from '../working_tree/views/Stack'
import { StackConfig } from '../working_tree/props/StackConfig'
import { StackAlignment } from '../working_tree/props/types'
import { Config } from '../working_tree/props/Config'
import { Renderer } from '../renderer/Renderer'
import { DummyViewManager } from '../renderer/DummyViewManager'

jest.useFakeTimers()
Renderer.setViewManager(new DummyViewManager())

function getRenderedTreeString() {
  return JSON.stringify(Renderer.getCurrentTree(), undefined, 2)
}

afterEach(() => {
  WorkingTree.dropAll()
  jest.setSystemTime(0)
})

test('Children of Stack(alignment=TopStart) get positioned correctly', () => {
  Stack(StackConfig('stack').width(400).height(400).alignment(StackAlignment.TopStart).padding(10), () => {
    Stack(Config('inner.1').width(200).height(200))
    Stack(Config('inner.2').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Stack(alignment=TopCenter) get positioned correctly', () => {
  Stack(StackConfig('stack').width(400).height(400).alignment(StackAlignment.TopCenter).padding(10), () => {
    Stack(Config('inner.1').width(200).height(200))
    Stack(Config('inner.2').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Stack(alignment=TopEnd) get positioned correctly', () => {
  Stack(StackConfig('stack').width(400).height(400).alignment(StackAlignment.TopEnd).padding(10), () => {
    Stack(Config('inner.1').width(200).height(200))
    Stack(Config('inner.2').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Stack(alignment=CenterStart) get positioned correctly', () => {
  Stack(StackConfig('stack').width(400).height(400).alignment(StackAlignment.CenterStart).padding(10), () => {
    Stack(Config('inner.1').width(200).height(200))
    Stack(Config('inner.2').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Stack(alignment=Center) get positioned correctly', () => {
  Stack(StackConfig('stack').width(400).height(400).alignment(StackAlignment.Center).padding(10), () => {
    Stack(Config('inner.1').width(200).height(200))
    Stack(Config('inner.2').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Stack(alignment=CenterEnd) get positioned correctly', () => {
  Stack(StackConfig('stack').width(400).height(400).alignment(StackAlignment.CenterEnd).padding(10), () => {
    Stack(Config('inner.1').width(200).height(200))
    Stack(Config('inner.2').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Stack(alignment=BottomStart) get positioned correctly', () => {
  Stack(StackConfig('stack').width(400).height(400).alignment(StackAlignment.BottomStart).padding(10), () => {
    Stack(Config('inner.1').width(200).height(200))
    Stack(Config('inner.2').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Stack(alignment=BottomCenter) get positioned correctly', () => {
  Stack(StackConfig('stack').width(400).height(400).alignment(StackAlignment.BottomCenter).padding(10), () => {
    Stack(Config('inner.1').width(200).height(200))
    Stack(Config('inner.2').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Stack(alignment=BottomEnd) get positioned correctly', () => {
  Stack(StackConfig('stack').width(400).height(400).alignment(StackAlignment.BottomEnd).padding(10), () => {
    Stack(Config('inner.1').width(200).height(200))
    Stack(Config('inner.2').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})
