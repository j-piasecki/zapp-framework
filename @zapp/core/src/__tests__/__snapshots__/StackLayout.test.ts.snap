// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Children of Stack(alignment=BottomCenter) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\",
    \\"isInherited\\": {}
  },
  \\"children\\": [
    {
      \\"id\\": \\"stack\\",
      \\"type\\": \\"stack\\",
      \\"config\\": {
        \\"id\\": \\"stack\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"stackAlignment\\": 7,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        },
        \\"isInherited\\": {}
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 200,
            \\"height\\": 200,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 200,
            \\"height\\": 200,
            \\"x\\": 100,
            \\"y\\": 190,
            \\"measured\\": true
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 100,
            \\"height\\": 100,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 150,
            \\"y\\": 290,
            \\"measured\\": true
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": true
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true
  }
}"
`;

exports[`Children of Stack(alignment=BottomEnd) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\",
    \\"isInherited\\": {}
  },
  \\"children\\": [
    {
      \\"id\\": \\"stack\\",
      \\"type\\": \\"stack\\",
      \\"config\\": {
        \\"id\\": \\"stack\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"stackAlignment\\": 8,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        },
        \\"isInherited\\": {}
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 200,
            \\"height\\": 200,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 200,
            \\"height\\": 200,
            \\"x\\": 190,
            \\"y\\": 190,
            \\"measured\\": true
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 100,
            \\"height\\": 100,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 290,
            \\"y\\": 290,
            \\"measured\\": true
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": true
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true
  }
}"
`;

exports[`Children of Stack(alignment=BottomStart) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\",
    \\"isInherited\\": {}
  },
  \\"children\\": [
    {
      \\"id\\": \\"stack\\",
      \\"type\\": \\"stack\\",
      \\"config\\": {
        \\"id\\": \\"stack\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"stackAlignment\\": 6,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        },
        \\"isInherited\\": {}
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 200,
            \\"height\\": 200,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 200,
            \\"height\\": 200,
            \\"x\\": 10,
            \\"y\\": 190,
            \\"measured\\": true
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 100,
            \\"height\\": 100,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 10,
            \\"y\\": 290,
            \\"measured\\": true
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": true
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true
  }
}"
`;

exports[`Children of Stack(alignment=Center) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\",
    \\"isInherited\\": {}
  },
  \\"children\\": [
    {
      \\"id\\": \\"stack\\",
      \\"type\\": \\"stack\\",
      \\"config\\": {
        \\"id\\": \\"stack\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"stackAlignment\\": 1,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        },
        \\"isInherited\\": {}
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 200,
            \\"height\\": 200,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 200,
            \\"height\\": 200,
            \\"x\\": 100,
            \\"y\\": 100,
            \\"measured\\": true
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 100,
            \\"height\\": 100,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 150,
            \\"y\\": 150,
            \\"measured\\": true
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": true
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true
  }
}"
`;

exports[`Children of Stack(alignment=CenterEnd) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\",
    \\"isInherited\\": {}
  },
  \\"children\\": [
    {
      \\"id\\": \\"stack\\",
      \\"type\\": \\"stack\\",
      \\"config\\": {
        \\"id\\": \\"stack\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"stackAlignment\\": 2,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        },
        \\"isInherited\\": {}
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 200,
            \\"height\\": 200,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 200,
            \\"height\\": 200,
            \\"x\\": 190,
            \\"y\\": 100,
            \\"measured\\": true
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 100,
            \\"height\\": 100,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 290,
            \\"y\\": 150,
            \\"measured\\": true
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": true
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true
  }
}"
`;

exports[`Children of Stack(alignment=CenterStart) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\",
    \\"isInherited\\": {}
  },
  \\"children\\": [
    {
      \\"id\\": \\"stack\\",
      \\"type\\": \\"stack\\",
      \\"config\\": {
        \\"id\\": \\"stack\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"stackAlignment\\": 0,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        },
        \\"isInherited\\": {}
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 200,
            \\"height\\": 200,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 200,
            \\"height\\": 200,
            \\"x\\": 10,
            \\"y\\": 100,
            \\"measured\\": true
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 100,
            \\"height\\": 100,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 10,
            \\"y\\": 150,
            \\"measured\\": true
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": true
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true
  }
}"
`;

exports[`Children of Stack(alignment=TopCenter) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\",
    \\"isInherited\\": {}
  },
  \\"children\\": [
    {
      \\"id\\": \\"stack\\",
      \\"type\\": \\"stack\\",
      \\"config\\": {
        \\"id\\": \\"stack\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"stackAlignment\\": 4,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        },
        \\"isInherited\\": {}
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 200,
            \\"height\\": 200,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 200,
            \\"height\\": 200,
            \\"x\\": 100,
            \\"y\\": 10,
            \\"measured\\": true
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 100,
            \\"height\\": 100,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 150,
            \\"y\\": 10,
            \\"measured\\": true
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": true
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true
  }
}"
`;

exports[`Children of Stack(alignment=TopEnd) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\",
    \\"isInherited\\": {}
  },
  \\"children\\": [
    {
      \\"id\\": \\"stack\\",
      \\"type\\": \\"stack\\",
      \\"config\\": {
        \\"id\\": \\"stack\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"stackAlignment\\": 5,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        },
        \\"isInherited\\": {}
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 200,
            \\"height\\": 200,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 200,
            \\"height\\": 200,
            \\"x\\": 190,
            \\"y\\": 10,
            \\"measured\\": true
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 100,
            \\"height\\": 100,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 290,
            \\"y\\": 10,
            \\"measured\\": true
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": true
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true
  }
}"
`;

exports[`Children of Stack(alignment=TopStart) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\",
    \\"isInherited\\": {}
  },
  \\"children\\": [
    {
      \\"id\\": \\"stack\\",
      \\"type\\": \\"stack\\",
      \\"config\\": {
        \\"id\\": \\"stack\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"stackAlignment\\": 3,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        },
        \\"isInherited\\": {}
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 200,
            \\"height\\": 200,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 200,
            \\"height\\": 200,
            \\"x\\": 10,
            \\"y\\": 10,
            \\"measured\\": true
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 100,
            \\"height\\": 100,
            \\"isInherited\\": {}
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 10,
            \\"y\\": 10,
            \\"measured\\": true
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": true
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true
  }
}"
`;
